### ISSUE TITLE
JS Directory Data File exported Promise not cached/singleton when used with Pagination Frontmatter

### ISSUE THREAD

**Issue by jonsage** [11ty/eleventy#707]
When using a Promise in a JS Data File and making a request to an external API, and using Pagination frontmatter to create multiple pages for each item in the API response, the API request happens multiple times (numbers of pages created + 1).

**Expected behavior**

I had thought subsequent calls to the same JS Data File would return the same Promise instance (and therefore not make multiple web requests).

**My code (simplified)**

**`docs/docs.11tydata.js`**
```javascript
module.exports = function () {
    return new Promise((resolve, reject) => {
        // logic to get data from web request
        
        // log each time this web request happens
        console.log('new web request');

        // resolve with web request response
        resolve({
            docs: [
                {
                    path: '',
                    content: '# five six'
                },
                {
                    path: 'one/two',
                    content: '*one two*'
                },
                {
                    path: 'three/four',
                    content: '**three four**'
                }
            ]
        });
    });
}
```

**`docs/index.njk`**
```markdown
---
permalink: /docs/{{ doc.path }}/
pagination:
  data: docs
  size: 1
  alias: doc
---

{{ doc.content | safe }}
```

**`console output`**
```
$ npx @11ty/eleventy
new web request
new web request
new web request
new web request
Writing _site/docs/index.html from ./docs/index.njk.
Writing _site/docs/one/two/index.html from ./docs/index.njk.
Writing _site/docs/three/four/index.html from ./docs/index.njk.
Processed 3 files in 0.32 seconds (v0.9.0)
```

**Environment:**
 - OS and Version: MacOS 10.14.6
 - Eleventy Version: 0.9.0

**Where in the code**
Here is the `require` that loads the JS Data File: 
 
https://github.com/11ty/eleventy/blob/19fe59856e373de04384fbcea2424ec9fe871434/src/TemplateData.js#L238

Could it be that because the `require` happens within an asynchronous function itself, that is the reason it ends up being called multiple times?

**Comment by jonsage** [comment]
For comparison - if I hack a new collection it works as expected - just one web request…

**`.eleventy.js`**
```javascript
module.exports = function (eleventyConfig) {
    eleventyConfig.addCollection("docs", function (collection) {
        return new Promise((resolve, reject) => {
            // logic to get data from web request
            
            // log each time this web request happens
            console.log('new web request');
    
            // resolve with web request response
            resolve([
                {
                    path: '',
                    content: '# five six'
                },
                {
                    path: 'one/two',
                    content: '*one two*'
                },
                {
                    path: 'three/four',
                    content: '**three four**'
                }
            ]);
        });
    });
}
```

**`docs/index.njk`**
```markdown
---
permalink: /docs/{{ doc.path }}/
pagination:
  data: collections.docs
  size: 1
  alias: doc
---

{{ doc.content | safe }}
```

**`console output`**
```
$ npx @11ty/eleventy
new web request
Writing _site/docs/index.html from ./docs/index.njk.
Writing _site/docs/one/two/index.html from ./docs/index.njk.
Writing _site/docs/three/four/index.html from ./docs/index.njk.
Processed 3 files in 0.25 seconds (v0.9.0)
```

**Comment by jonsage** [comment]
After reading, and learning that node modules are singletons/cached by default (https://stackoverflow.com/a/38874910) I tried exporting the promise directly, and not the function (in the first example)…

**`docs/docs.11tydata.js`**
```javascript
module.export = new Promise((resolve, reject) => {
    // logic to get data from web request
    
    // log each time this web request happens
    console.log('new web request');

    // resolve with web request response
    resolve({
        docs: [
            {
                path: '',
                content: '# five six'
            },
            {
                path: 'one/two',
                content: '*one two*'
            },
            {
                path: 'three/four',
                content: '**three four**'
            }
        ]
    });
});
```

But this causes an error:

```
> Could not resolve pagination key in template data: docs

`Error` was thrown:
    Error: Could not resolve pagination key in template data: docs
    …
```

**Comment by Ryuno-Ki** [comment]
I work around it by making the API request outside of eleventy (writing to disk) and read its content inside a `_data` file.
However this is also async (shouldn't matter whether via the wire or disk).

I return a function which returns a promise ... like your case 1.
So something's different :-D

Let me know whether I shall some code.

**Comment by Ryuno-Ki** [comment]
+ share

**Comment by jonsage** [comment]
@Ryuno-Ki after reading what you say I tried using a global `_data/docs.js` file instead of one within a specific directory `docs/docs.11tydata.js` and it appears the problem is with the data file being in the directory… **when it's in the global `_data` directory the request runs once as expected.**

So the combination of factors seem to be the **frontmatter paging of a directory data file** - in this case the problem arises.

My previous assertion that it was _"number of pages + 1"_ also appears incorrect, as today my first example is logging even more (although I am using a newer version of node)… I think that might be a red herring.

**Comment by Ryuno-Ki** [comment]
Cool, so we could at least narrow it down!
Would it make sense to update the title?

**Comment by zachleat** [comment]
Sorry for the super late response here but I believe this is a duplicate of #919, fixed in 1.0.0!

> This is an automated message to let you know that a helpful response was posted to your issue and for the health of the repository issue tracker the issue will be closed. This is to help alleviate issues hanging open waiting for a response from the original poster.

> If the response works to solve your problem—great! But if you’re still having problems, do not let the issue’s closing deter you if you have additional questions! Post another comment and we will reopen the issue. Thanks!
